nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/RunMongoAggregation.java

```diff
@@ -0,0 +1,180 @@
+
+    static final PropertyDescriptor QUERY = new PropertyDescriptor.Builder()
+            .name("mongo-agg-query")
+            .displayName("Query")
+            .expressionLanguageSupported(true)
+            .description("The aggregation query to be executed.")
+            .required(true)
+            .addValidator(AGG_VALIDATOR)
+            .build();
+
+    static {
+        List<PropertyDescriptor> _propertyDescriptors = new ArrayList<>();
+        _propertyDescriptors.addAll(descriptors);
+        _propertyDescriptors.add(QUERY);
+        _propertyDescriptors.add(BATCH_SIZE);
+        _propertyDescriptors.add(RESULTS_PER_FLOWFILE);
+        _propertyDescriptors.add(SSL_CONTEXT_SERVICE);
+        _propertyDescriptors.add(CLIENT_AUTH);
+        propertyDescriptors = Collections.unmodifiableList(_propertyDescriptors);
+
+        final Set<Relationship> _relationships = new HashSet<>();
+        _relationships.add(REL_SUCCESS);
+        _relationships.add(REL_ORIGINAL);
+        relationships = Collections.unmodifiableSet(_relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return propertyDescriptors;
+    }
+
+    static String buildBatch(List batch) {
+        ObjectMapper mapper = new ObjectMapper();
+        String retVal;
+        try {
+            retVal = mapper.writeValueAsString(batch.size() > 1 ? batch : batch.get(0));
+        } catch (Exception e) {
+            retVal = null;
+        }
+
+        return retVal;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+
+        String query = context.getProperty(QUERY).evaluateAttributeExpressions(flowFile).getValue();
+        Integer batchSize = context.getProperty(BATCH_SIZE).asInteger();
+        Integer resultsPerFlowfile = context.getProperty(RESULTS_PER_FLOWFILE).asInteger();
+
+        MongoCollection collection = getCollection(context);
+        MongoCursor iter = null;
+
+        try {
+            List<Bson> aggQuery = buildAggregationQuery(query);
+            AggregateIterable it = collection.aggregate(aggQuery);
+            it.batchSize(batchSize != null ? batchSize : 1);
+
+            iter = it.iterator();
+            List batch = new ArrayList();
+
+            while (iter.hasNext()) {
+                batch.add(iter.next());
+                if (batch.size() == resultsPerFlowfile) {
+                    writeBatch(buildBatch(batch), context, session);
+                    batch = new ArrayList();
+                }
+            }
+
+            if (batch.size() > 0) {
+                writeBatch(buildBatch(batch), context, session);
+            }
+
+            session.transfer(flowFile, REL_ORIGINAL);
+        } catch (IOException i) {
+            context.yield();
```

Comment:
If there is an input flow file, an error should be logged, and the flow file transferred to a different relationship, rather than rollback. Depending on the kinds of errors that can occur, you may want a "retry" relationship (along with possible penalization) if the same flow file might succeed later (due to connection issues, e.g.) or "failure" if the flow file will always fail due to bad data, format, etc.

If there is no input flow file, a yield and rollback should be ok I think.